sisl/viz/data_sources/data_source.py

```diff
@@ -0,0 +1,98 @@
+class DataSource:
+
+    def _get_composite_source(self, op, other=None):
+        """Given an operator, returns a DataSource that will evaluate to it."""
+        return CompositeDataSource(self, op, other)
+
+    def _get_unary_source(self, op):
+        """Given an operator, returns a DataSource that will evaluate to it."""
+        return UnaryDataSource(self, op)
+
+    # Define all supported operators
```

Comment:
In the `DataSource` class I want to implement all operators like:

```python
def __op__(self, other):
    return self._get_composite_source("__op__", other)
```

I've tried setting all attributes dynamically in a loop:
```python
class A:
    pass

ops = ("__mul__", "__add__")

for op in ("__mul__", "__add__"):
    def function(self, other):
        print(op)
    setattr(A, op, function)
```
But then `op` keeps changing so every method gets the last `op`, and
```python
A() * 2
```
prints `__add__`. Do you have any idea how can I acheive this?

Reply:
I think sadly that in this case it is because the `op` is *global*.
The fix is to hide things in a wrapper function, then the variable becomes a local one:
```python
ops = ("__mul__", "__add__")
for op in ops:
    def wrapper(op):
        def function(self, other):
            r""" {op} """
            print(f"{self.__class__.__name__}", op, other)
        return function
    setattr(A, op, wrapper(op))
```
It ain't pretty, but it works.

Reply:
It seems to be because `op` is the *same* globally defined variable, I tried to change variable to a 2nd one to hide it, to no avail. Agreed this is a bit weird... ;(

Reply:
> I tried to change variable to a 2nd one to hide it, to no avail.

Yeah tried the same thing :sweat_smile:

It ain't pretty, but much better than defining all of them explicitly! Thanks!

Also, basically I want to support all operations that numpy arrays support, do you know if there's a better way instead of defining a list of operations?

Reply:
Niceee! This has simplified the code a lot :)

Reply:
And it is much better, since you can do `np.log( BondLength() )` and it will get evaluated lazily as well. This is great!

Reply:
Exactly ;)

Please note my comment in your commit, `numpy.ufuncs` can do lots of things that might break behaviour if not carefully implemented. :)